### PR TITLE
Add `ImageOptions` to artifact builder.

### PR DESCRIPTION
**Pull request by gsquared94**
Fixes: #4370  

**Description**

This PR introduces `ImageOptions` to the different builders and currently only partially implements it for docker builder and buildpacks.
It recognizes `Dev` and `Debug` as build configurations and passes different build args like `GO_GCFLAGS` when building the artifact for debug vs dev. It also updates hashcode formula for image caching so that images for `Dev` vs `Debug` are cached separately.

**User facing changes (remove if N/A)**

Dockerfile can expose `ARG SKAFFOLD_GO_GCFLAGS` and `ARG SKAFFOLD_GO_LDFLAGS` and expect Skaffold to supply meaningful values.
Buildpacks auto set `GOOGLE_GCFLAGS` and `GOOGLE_LDFLAGS` (so far).

Note: _The artifact cached during `skaffold dev` will have a different hashcode than `skaffold debug` for all builders going forward_

### CODE REVIEW

File: integration/examples/getting-started/Dockerfile [hunk 1]
```diff
@@ -1,6 +1,11 @@
 FROM golang:1.12.9-alpine3.10 as builder
+
+# These flag values are provided when building for debug
+ARG GO_GCFLAGS
```
Comment (gsquared94) [inline integration/examples/getting-started/Dockerfile:4]: `GO_GCFLAGS` and `GO_LDFLAGS` are explicitly added as dockerfile args and their values are provided by Skaffold

File: integration/testdata/debug/go/Dockerfile [hunk 2]
```diff
@@ -2,8 +2,8 @@ FROM golang:1.12 as builder
 
 COPY app.go .
 # Must use eval to handle GOGCFLAGS with spaces like `-gcflags='all=-N -l'`
-ARG GOGCFLAGS
-RUN eval go build "${GOGCFLAGS}" -o /app .
+ARG GO_GCFLAGS
```
Comment (gsquared94) [inline integration/testdata/debug/go/Dockerfile:5]: revert this change.

File: integration/testdata/debug/go/Dockerfile [hunk 3]
```diff
@@ -2,8 +2,8 @@ FROM golang:1.12 as builder
...
+RUN eval go build "${GO_GCFLAGS}" -o /app .
```
Comment (gsquared94) [inline integration/testdata/debug/go/Dockerfile:6]: fix command `--gcflags="${GO_GCFLAGS}"`

File: pkg/skaffold/build/cache/hash.go [hunk 4]
```diff
@@ -89,6 +90,15 @@ func getHashForArtifact(ctx context.Context, depLister DependencyLister, a *late
 		inputs = append(inputs, evaluatedEnv...)
 	}
 
+	// add image options hash
```
Comment (gsquared94) [inline pkg/skaffold/build/cache/hash.go:93]: cache takes `opts` hash value

File: pkg/skaffold/build/cluster/cluster.go [hunk 5]
```diff
@@ -48,13 +48,13 @@ func (b *Builder) Build(ctx context.Context, out io.Writer, tags tag.ImageTags,
 	return build.InParallel(ctx, out, tags, artifacts, b.buildArtifact, b.ClusterDetails.Concurrency)
 }
 
-func (b *Builder) buildArtifact(ctx context.Context, out io.Writer, artifact *latest.Artifact, tag string) (string, error) {
-	digest, err := b.runBuildForArtifact(ctx, out, artifact, tag)
+func (b *Builder) buildArtifact(ctx context.Context, out io.Writer, artifact *latest.Artifact, opts *build.ImageOptions) (string, error) {
```
Comment (gsquared94) [inline pkg/skaffold/build/cluster/cluster.go:51]: all builders accept `build.ImageOptions`

File: pkg/skaffold/build/options.go [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+// Configuration denotes build configuration for the artifact builder as either Release or Debug
+type Configuration int
+
+const (
+	Release = iota
+	Debug
+)
+
+// ImageOptions provides options for the artifact builder
+type ImageOptions struct {
```
Comment (gsquared94) [inline pkg/skaffold/build/options.go:37]: `ImageOptions` encapsulates image tag along with the current configuration (Debug or Release) and a map of additional properties

Comment (gsquared94) [inline pkg/skaffold/build/options.go:37]: `ImageOptions` encapsulates image tag along with the current configuration (Debug or Release) and a map of additional properties

File: pkg/skaffold/docker/image.go [hunk 7]
```diff
@@ -190,7 +190,10 @@ func (l *localDaemon) Build(ctx context.Context, out io.Writer, workspace string
 		ForceRemove: l.forceRemove,
 		NetworkMode: strings.ToLower(a.NetworkMode),
 		NoCache:     a.NoCache,
-	})
+	}
+
+	imageOpts = opts.ApplyModifiers(imageOpts)
```
Comment (gsquared94) [inline pkg/skaffold/docker/image.go:195]: Applying docker builder options

File: pkg/skaffold/docker/options.go [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package docker
+
+import "github.com/docker/docker/api/types"
```
Comment (gsquared94) [inline pkg/skaffold/docker/options.go:19]: Docker builder options implementation

File: pkg/skaffold/build/options.go [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+}
+
+var CurrentConfiguration Configuration
+
+func CreateBuilderOptions(tag string) *ImageOptions {
+	return &ImageOptions{
+		Tag:           tag,
+		Configuration: CurrentConfiguration,
+	}
+}
```
Comment (briandealwis) [inline pkg/skaffold/build/options.go:51]: We really should avoid global state.  [`SkaffoldOptions.Command`](https://github.com/GoogleContainerTools/skaffold/blob/master/pkg/skaffold/config/options.go#L76) records the command used on the command-line, and offers [`IsDevMode()` and `IsDebugMode()`](https://github.com/GoogleContainerTools/skaffold/blob/master/pkg/skaffold/config/options.go#L92-L98) to test.

I wonder if it would be worth broadening the scope of this structure to capture the build options, including the artifact, and have it be passed in and around.

File: pkg/skaffold/build/local/local.go [hunk 10]
```diff
@@ -112,3 +113,11 @@ func (b *Builder) getImageIDForTag(ctx context.Context, tag string) (string, err
 	}
 	return insp.ID, nil
 }
+
+func ToDockerOpts(opts *build.ImageOptions) *docker.BuildOptions {
```
Comment (briandealwis) [inline pkg/skaffold/build/local/local.go:117]: This should be done in the docker builder itself.

File: pkg/skaffold/build/local/local.go [hunk 11]
```diff
@@ -86,19 +87,19 @@ func (b *Builder) runBuildForArtifact(ctx context.Context, out io.Writer, a *lat
 
 	switch {
 	case a.DockerArtifact != nil:
-		return b.buildDocker(ctx, out, a, tag)
+		return b.buildDocker(ctx, out, a, ToDockerOpts(opts))
```
Comment (briandealwis) [inline pkg/skaffold/build/local/local.go:90]: It feels like this ImageOptions should be passed down into the builders.

File: pkg/skaffold/build/local/local.go [hunk 12]
```diff
@@ -45,8 +46,8 @@ func (b *Builder) Build(ctx context.Context, out io.Writer, tags tag.ImageTags,
 	return build.InParallel(ctx, out, tags, artifacts, b.buildArtifact, *b.cfg.Concurrency)
 }
 
-func (b *Builder) buildArtifact(ctx context.Context, out io.Writer, a *latest.Artifact, tag string) (string, error) {
-	digestOrImageID, err := b.runBuildForArtifact(ctx, out, a, tag)
+func (b *Builder) buildArtifact(ctx context.Context, out io.Writer, a *latest.Artifact, opts *build.ImageOptions) (string, error) {
```
Comment (briandealwis) [inline pkg/skaffold/build/local/local.go:49]: Since we're passing in the destination image tag, opts shouldn't be a pointer.

File: pkg/skaffold/docker/options.go [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+
+var buildArgsForDebug = map[string]string{
+	"GO_GCFLAGS": "'all=-N -l'", // disable build optimization for Golang
```
Comment (briandealwis) [inline pkg/skaffold/docker/options.go:22]: What do you think about prefixing these with `SKAFFOLD_`?

File: pkg/skaffold/build/options.go (hunk 6 above)
Comment (gsquared94) [reply pkg/skaffold/build/options.go:37]: Since it's in package `build` struct names shouldn't start with `Build-`(at least the IDE prompts so 🤷🏾‍♂️) which is why I named it so; but I'll think of something else.

File: pkg/skaffold/build/options.go [hunk 14]
```diff
@@ -0,0 +1,85 @@
...
+	"fmt"
+	"sort"
+	"strconv"
+)
+
+// Configuration denotes build configuration for the artifact builder as either Release or Debug
+type Configuration int
+
+const (
+	Release = iota
```
[the remark this answers is not in the document]
Comment (gsquared94) [reply pkg/skaffold/build/options.go]: I've changed this to only have `Dev` and `Debug` since I also use this parameter in calculating cache hashcode. So having separate `Build` type would cause the image built by `skaffold build` not to be used for next `skaffold dev`

File: pkg/skaffold/build/local/local.go (hunk 10 above)
Comment (gsquared94) [reply pkg/skaffold/build/local/local.go]: Pushing `ImageOptions` into any of the actual builder implementations causes an import cycle (you had predicted this!). So I've defined `ImageOptions` as the generic builder-agnostic options object. The different builders define their own options object in their respective adapter package.

File: pkg/skaffold/build/local/local.go (hunk 11 above)
Comment (gsquared94) [reply pkg/skaffold/build/local/local.go:89]: > Copied from above

Pushing ImageOptions into any of the actual builder implementations causes an import cycle (you had predicted this!). So I've defined ImageOptions as the generic builder-agnostic options object. The different builders define their own options object in their respective adapter package.

File: pkg/skaffold/build/local/local.go (hunk 12 above)
Comment (gsquared94) [reply pkg/skaffold/build/local/local.go]: done

File: pkg/skaffold/docker/options.go (hunk 13 above)
Comment (gsquared94) [reply pkg/skaffold/docker/options.go]: done

File: pkg/skaffold/build/options.go (hunk 9 above)
Comment (gsquared94) [reply pkg/skaffold/build/options.go]: Done

File: pkg/skaffold/build/options.go [hunk 15]
```diff
@@ -0,0 +1,85 @@
...
+	Debug
+)
+
+// ImageOptions provides options for the artifact builder
+type ImageOptions struct {
+	// FIXME: Tag should be []string but we don't support multiple tags yet
+	Tag           string             // image tag
+	Configuration Configuration      // build image for release or debug
+	Args          map[string]*string // additional builder specific args
```
[the remark this answers is not in the document]
Comment (gsquared94) [reply pkg/skaffold/build/options.go]: I've renamed it to `BuildArgs`. 
The way I see this being used is in differentiating parameters that could influence build between the different skaffold actions like Dev, Build or Debug. Since skaffold.yaml doesn't have sections for these in case there are different settings required those can be put in this generic map. Whichever builder understands that setting would utilize it. This way since the setting affects the built image it also participates in cache hashcode calculation. WDYT?

File: pkg/skaffold/build/options.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (gsquared94) [reply pkg/skaffold/build/options.go]: I've changed this to only have Dev and Debug since I also use this parameter in calculating cache hashcode. So having separate Build type would cause the image built by skaffold build not to be used for next skaffold dev

File: pkg/skaffold/build/options.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (gsquared94) [reply pkg/skaffold/build/options.go]: A concrete example could be to push the buildpacks builder image digest to this map so that if that changes it invalidates the cache. That'd solve this TODO:
image

File: integration/testdata/debug/go/Dockerfile [hunk 16]
```diff
@@ -1,9 +1,10 @@
 FROM golang:1.12 as builder
 
 COPY app.go .
-# Must use eval to handle GOGCFLAGS with spaces like `-gcflags='all=-N -l'`
-ARG GOGCFLAGS
-RUN eval go build "${GOGCFLAGS}" -o /app .
+ARG SKAFFOLD_GO_GCFLAGS
+ARG SKAFFOLD_LD_GCFLAGS
```
[the remark this answers is not in the document]
Comment (gsquared94) [reply integration/testdata/debug/go/Dockerfile:5]: 😅 🤦🏽‍♂️

**Comment by gsquared94** [comment]
Closing in favor of #4606
